README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# jpndistrict <img src="man/figures/logo.png" align="right" width="120px" />

[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/jpndistrict)](https://cran.r-project.org/package=jpndistrict)
[![Coverage
status](https://codecov.io/gh/uribo/jpndistrict/branch/master/graph/badge.svg)](https://codecov.io/github/uribo/jpndistrict?branch=master)

## Overview

In this package, the administrative area data to be provided uses the
National Land Numerical Information. Shinya Uryu is editing and
processing this data. Therefore, when preparing a secondary work using
this data, it is necessary to follow the term of the National Land
Numerical Information.

This package provide map data is based on the Digital Map 25000 (Map
Image) published by Geospatial Information Authority of Japan (Approval
No.603FY2017 information usage <https://www.gsi.go.jp>).

## Installation

Install from CRAN.

``` r
install.packages("jpndistrict")
```

For developers, please use the **devtool** package to install via the
GitHub repository.

``` r
install.packages("remotes")
remotes::install_github("uribo/jpndistrict")
```

## Usage

``` r
# Load Package
library(jpndistrict)
```

### Administrative area data

``` r
jpn_pref(14)
jpn_pref(14, district = FALSE)
jpn_cities(14, admin_name = "海老名市")
jpn_cities(33, admin_name = c("倉敷市", "笠岡市"))
```

``` r
# Return mesh polygons included in administrative areas.
mesh_district(jis_code = "05")
mesh_district(jis_code = 33101)
```

### Administration office data

``` r
jpn_admins(jis_code = 33)
jpn_admins(jis_code = c("33101", "33212"))
```

### Reverse geocoding in prefecture and city level

``` r
find_pref(longitude = 133.915, latitude = 34.666)
find_city(longitude = 133.915, latitude = 34.666)
# sfg
find_city(geometry = sf::st_point(c(140.112, 36.083)))
```

### City name and code validation

``` r
code_validate(jis_code = 33101)
code_reform(jis_code = c(1, "33", "08201"))
```


NEWS.md
# jpndistrict (development version)

# jpndistrict 0.3.9

# jpndistrict 0.3.8 (2020-11-27)

- Using memoise

# jpndistrict 0.3.7 (2020-06-12)

- Reviewing how to download files, use curl from now on.

# jpndistrict 0.3.6 (2020-04-20)

- Changes to accomodate sf: st_make_valid no longer in lwgeom ([#32](https://github.com/uribo/jpndistrict/issues/32)).

# jpndistrict 0.3.5 (2020-04-02)

- Follow up sf v0.9.0 update.
- The geometry of the dataset referenced inside the package was encoded by Google's Polyline Encoding Algorithm with googlePolylines. This has reduced the file size.

# jpndistrict 0.3.4

- Fix CRAN package check results with Latin-1 error.
- Refactoring of dependency packages and functions mainly using rlang and tidyverse.
- Renewal pkgdown site and travis ci environment with tic.

# jpndistrict 0.3.3

## API

- Added new functions to validate jis code. (`code_validate()`, `code_reform()` and `cityname_reform()`)
- English column names have been added to the `jpnprefs` data set. [#21](https://github.com/uribo/jpndistrict/pull/21), by Ryo Nakagawara (@Ryo-N7)
- Added new function to find city code (`find_jis_code()`). This function do not consider historical changes. [#23](https://github.com/uribo/jpndistrict/pull/23)

## Fixed Issues

- Mismatching variable names in `jpn_pref()` [#17](https://github.com/uribo/jpndistrict/issues/17)
- Always returns the nearest city or prefecture [#18](https://github.com/uribo/jpndistrict/issues/18)
- Failed in `code_validate()` [#19](https://github.com/uribo/jpndistrict/issues/19)
- Unify behavior due to differences in parameters in `jpn_pref()`. [#22](https://github.com/uribo/jpndistrict/issues/22)

# jpndistrict 0.3.2

- Clarified the arguments that to select administration area. pref_code is jis code from 1 to 47. jis_code is an identifier for a city that range of 5 digits.
- `find_city()` is support sf geometry.

## Fixed Issues

- Remove redundant character from prefecture code 13 data [#13](https://github.com/uribo/jpndistrict/issues/13)
- `mesh_district()` now return only selected administation area [#15](https://github.com/uribo/jpndistrict/issues/15)
- Broken `district_viewer()` [#16](https://github.com/uribo/jpndistrict/issues/16)


# jpndistrict 0.3.1

- Setup tidy development environment.

## Issues

- Fix some bugs.
    - City administration data update [#4](https://github.com/uribo/jpndistrict/issues/4)
    - Reverse geocoding failed behabior [#10](https://github.com/uribo/jpndistrict/issues/10)

# jpndistrict 0.3.0

- Updated citation and licence. 
    - Approval No.603FY2017 information usage

## Issues

- Fix some bugs ([#6](https://github.com/uribo/jpndistrict/issues/6), [#7](https://github.com/uribo/jpndistrict/issues/7)).

## API

- New function a find from coordinates to prefecture and city that `find_pref()` and `find_city()`.
- Add a function `mesh_district()` that export to meshed city area.
- Projected coordinates are unified to WGS84 (espg: 4326).
- Rename `spdf_jpn_pref()` function to `jpn_pref()`.

# jpndistrict 0.2.0

- Added support for sf.
- Remove `gg_jpn_district()`. Use instead `ggplot2::geom_sf()` (ggplot2 > 2.2.1)
- Convert string variables to factor always FALSE.

# jpndistrict 0.1.1

- Modified raw data download method `wget` to `auto`.
- Character encoding for administration area dataset to UTF8. ([#2](https://github.com/uribo/jpndistrict/issues/2))

# jpndistrict 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* Prepare base functions.
* Added `district_viewer()` for shiny gadgets.


R/code_validate.R
#' Administration code validation
#'
#' @param jis_code jis code for prefecture and city identifical number.
#' If prefecture, must be from 1 to 47. If city, range of 5 digits.
#' @importFrom purrr flatten_chr keep map map_chr map_if
#' @note The `code_validate` function was added in version 0.3.2.9000
#' @examples
#' code_validate(jis_code = "05")
#' code_validate(jis_code = 33101)
#' code_validate(jis_code = c("01", "33101"))
#' @export
code_validate <- function(jis_code) {
  . <- NULL # nolint
  res <-
    code_reform(jis_code)
  names(res) <-
    res %>%
    purrr::map(nchar) %>%
    purrr::map_chr(~ ifelse(.x == 2, "prefecture", "city"))
  res <-
    res %>%
    purrr::map_if(names(.) == "prefecture", ~ prefcode_validate(.x)) %>%
    purrr::map_if(names(.) == "city", ~ match_city_name(.x)$city_code) %>%
    purrr::keep(~ length(.x) == 1)
  list(administration_type = names(res),
       code = purrr::flatten_chr(res))
}

#' Reform input jis code as 2 or 5 character length.
#' @inheritParams code_validate
#' @note The `code_reform` function was added in version 0.3.2.9000
#' @name code_reform
#' @examples
#' code_reform(c(1, "33", "08201"))
#' @export
code_reform <- function(jis_code) {
  . <- NULL # nolint
  checked <-
    jis_code %>%
    purrr::map(nchar) %>%
    purrr::keep(~ .x %in% c(1, 2, 5)) %>%
    length()
  if (length(jis_code) != checked)
    rlang::abort("Input jis-code must to 2 or 5 digits.")
  jis_code %>%
    purrr::map(as.numeric) %>%
    purrr::map_if(.p = nchar(.) %in% c(1, 2), ~ sprintf("%02d", .x)) %>%
    purrr::map_if(.p = nchar(.) %in% c(4, 5), ~ sprintf("%05d", .x)) %>%
    purrr::flatten_chr()
}

prefcode_validate <- function(pref_code) {
  codes <-
    sapply(seq(1, 47, 1), sprintf, fmt = "%02d")
  if (identical(codes[codes %in% pref_code], character(0)))
    rlang::abort("jis_code must be start a integer or as character from 1 to 47.") # nolint
  pref_code
}

match_city_name <- function(jis_code) {
  city_code <- NULL
  df <-
    code_reform(jis_code) %>%
    purrr::map_chr(~ substr(.x, 1, 2)) %>%
    unique() %>%
    purrr::map_dfr(
      ~ jpn_pref(.x, district = TRUE) %>%
        sf::st_drop_geometry() %>%
        dplyr::select(city_code, city))
  res <-
    subset(df, city_code %in% jis_code)
  n_mismatch <-
    length(jis_code[!jis_code %in% res$city_code])
  if (n_mismatch >= 1)
    rlang::inform(
      paste(n_mismatch, "matching code were not found."))
  res
}
